packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts

```diff
@@ -297,12 +300,14 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
       internetFacing
     };
 
-    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer : new ApplicationLoadBalancer(this, 'LB', lbProps);
+    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer as ApplicationLoadBalancer
```

Comment:
Wait, this is dangerous. Why is the input argument changed to `IApplicationLoadBalancer`, but it's downcast to a concrete class here?

Reply:
To completely explain: this cast is not necessarily correct. For example, for an imported load balancer this cast is DEFINITELY NOT correct, as an `ApplicationLoadBalancer` is going to have methods that an `ImportedApplicationLoadBalancer` does not have, and if you try to use those you will get crashes.

If the problem is that `this.loadBalancer` has been typed as an `ApplicationLoadBalancer` and so you must trick the type system, this is not the way to do it! A way to "properly" handle this case while not breaking backwards compatibility would be something like this:

```ts
class {
  private readonly _applicationLoadBalancer?: ApplicationLoadBalancer;

  constructor() {
    if (props.loadBalancer instanceof ApplicationLoadBalancer) {
      this._applicationLoadBalancer = props.loadBalancer;
    }
  } 

  public get applicationLoadBalancer(): ApplicationLoadBalancer {
    if (!this._applicationLoadBalancer) {
      throw new Error('.applicationLoadBalancer can only accessed if the class was constructed with an owned load balancer, not an imported one');
    } 
    return this._applicationLoadBalancer;
  }
}
```

That is, the type-safe API promises things we don't want to guarantee anymore, so now we have to transform it into a run-time check.

Not ideal, but the best we can do given the contracts we uphold.

Reply:
Thanks for the detailed feedback, this was really helpful in understanding the guards we put against backwards compatibility.